src/Oscoin/Data/OscoinTx.hs

```diff
@@ -0,0 +1,322 @@
+
+applyTxPayload (TxUnregisterProject addr) author ws = do
+    p@Project{..} <- lookupProject addr   ws
+    member        <- lookupMember  author ws
+
+    handler (pUnregister p member)
+    pure (WorldState.delete projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+applyTxPayload (TxAuthorize addr pk) author ws = do
+    p@Project{..} <- lookupProject addr    ws
+    member        <- lookupMember  author  ws
+    epoch         <- lookupEpoch           ws
+
+    handler (pAuthorize pk p member)
+    pure (adjust (addKey epoch) projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+    addKey e v = case v of
+        ProjectVal p -> ProjectVal $
+            p { pMaintainers = Map.insert pk (mkMember pk e) (pMaintainers p) }
+        _ -> v
+
+applyTxPayload (TxDeauthorize addr pk) author ws = do
+    p@Project{..} <- lookupProject addr    ws
+    member        <- lookupMember  author  ws
+
+    handler (pDeauthorize pk p member)
+    pure (adjust removeKey projKey ws, TxOutput)
+  where
+    projKey = addressKey addr
+
+    removeKey v = case v of
+        ProjectVal p -> ProjectVal $
+            p { pMaintainers = Map.delete pk (pMaintainers p) }
+        _ -> v
+
+applyTxPayload TxCheckpoint{} _ _ws = notImplemented
+applyTxPayload TxUpdateContract{} _ _ws = notImplemented
+
+applyTxPayload (TxTransfer _ _ bal) _ _ | bal == 0 =
+    Left (ErrInvalidTransfer bal)
+applyTxPayload (TxTransfer a b bal) author ws = do
+    acc <- lookupAccount a ws
+
+    let balance = accountBalance acc
+        in if author == a
+            then if balance > bal
+                 then Right (transfer, TxOutput)
+                 else Left (ErrInsufficientBalance balance)
+            else Left (ErrNotAuthorized author)
+
+  where
+    aKey = addressKey a
+    bKey = addressKey b
+
+    transfer =
+        alter (withAccount (withdraw bal)) aKey .
+        alter (withAccount (deposit bal))  bKey $ ws
+
+    withAccount f v = case v of
+        -- Account exists. Update balance or delete account if zero balance.
+        Just (AccountVal acc) -> case f acc of
+            acc'@Account{..} | accountBalance == 0 -> Nothing
+                             | otherwise           -> Just (AccountVal acc')
+        -- Account doesn't exist. Create it.
+        Nothing -> Just . AccountVal $ (mkAccount b) { accountBalance = bal }
+
+        -- Not an account ...
+        val -> val
+
+-- | Map a handler result into a transaction result.
+handler :: Either Error a -> Either (TxError c) a
+handler (Left err) = Left (ErrHandlerFailed err)
+handler (Right r)  = Right r
+
+-- | Lookup a member in the state.
+lookupMember
+    :: (ByteArrayAccess (Crypto.ShortHash c), CBOR.Serialise (Crypto.ShortHash c))
+    => Address c -> WorldState c -> Either (TxError c) (Member c)
+lookupMember addr ws =
+    case WorldState.lookup (addressKey addr) ws of
+        Just (MemberVal m) -> Right m
+        _                  -> Left (ErrAccountNotFound addr)
```

Comment:
It is a bit unsatisfying that the type of the key doesn't correspond to the value - one can store anything under an `Address`, and later try to get back a different thing using the same key. While this can be solved using singleton types, wouldn't it be more informative in the meantime to map the absence of a key to a single `ErrNotFound`, and signal the case of a type mismatch as a separate error constructor?

Reply:
Good idea - something like this 7bd9b4f ? The error could be improved in the future to include expected/actual..